### PR TITLE
Feature/data tree diff

### PR DESCRIPTION
**Pull request by annikahannig**
Hey everyone!

My colleagues at DE-CIX require the `lyd_diff` function from the C library in a python project.
I added said function and integrated it into the library.

Hope you like it,

Anni

### CODE REVIEW

File: cffi/cdefs.h [hunk 1]
```diff
@@ -434,6 +437,21 @@ struct lyd_node_leaf_list {
 	...;
 };
 
+typedef enum {
+    LYD_DIFF_END = 0,
```
Comment (rjarry) [inline cffi/cdefs.h:441]: Instead of specifying the value here, you should use the `...` ellipsis after the last `enum` value you want to export to let CFFI use the values from `libyang.h`. That way, if they change the values, the extension will follow automatically.

File: cffi/cdefs.h [hunk 2]
```diff
@@ -434,6 +437,21 @@ struct lyd_node_leaf_list {
...
+    LYD_DIFF_CHANGED,
+    LYD_DIFF_MOVEDAFTER1,
+    LYD_DIFF_CREATED,
+    LYD_DIFF_MOVEDAFTER2
+} LYD_DIFFTYPE;
+
+struct lyd_difflist {
+    LYD_DIFFTYPE *type;
+    struct lyd_node **first;
+    struct lyd_node **second;
```
Comment (rjarry) [inline cffi/cdefs.h:452]: Maybe add an ellipsis `...` here as well, just in case new fields are added to `struct lyd_difflist`.

File: libyang/data.py [hunk 3]
```diff
@@ -224,6 +281,31 @@ def validate(
...
+        self,
+        other: "DNode",
+        no_siblings: bool = False,
+        with_defaults: bool = False,
+    ) -> Iterator[DDiff]:
+        flags = diff_flags(no_siblings=no_siblings, with_defaults=with_defaults)
+        difflist = lib.lyd_diff(self.cdata, other.cdata, flags)
+
+        i = 0
+        while True:
```
Comment (rjarry) [inline libyang/data.py:294]: I'm not fond of `while True` constructs when it can be done differently.

Could you rework this like the following?

```python3
while difflist.type[i] != lib.LYD_DIFF_END:
```

File: libyang/data.py [hunk 4]
```diff
@@ -224,6 +281,31 @@ def validate(
...
             raise self.context.error("validation failed")
 
+    def diff(
+        self,
+        other: "DNode",
+        no_siblings: bool = False,
+        with_defaults: bool = False,
+    ) -> Iterator[DDiff]:
+        flags = diff_flags(no_siblings=no_siblings, with_defaults=with_defaults)
+        difflist = lib.lyd_diff(self.cdata, other.cdata, flags)
```
Comment (rjarry) [inline libyang/data.py:291]: You should check that difflist is not `NULL`. Something like:

```python3
if not difflist:
    raise self.context.error("diff failed")
```

File: libyang/data.py [hunk 5]
```diff
@@ -224,6 +281,31 @@ def validate(
...
+        while True:
+            dtype = difflist.type[i]
+            first = None
+            second = None
+            if dtype == lib.LYD_DIFF_END:
+                break
+            if difflist.first[i] != ffi.NULL:
+                first = self.new(self.context, difflist.first[i])
+            if difflist.second[i] != ffi.NULL:
+                second = self.new(self.context, difflist.second[i])
```
Comment (rjarry) [inline libyang/data.py:303]: I prefer to avoid comparing values to `ffi.NULL`. Could you rewrite this like:

```python3
first = self.new(self.context, difflist.first[i]) if difflist.first[i] else None
second = self.new(self.context, difflist.second[i]) if difflist.second[i] else None
```

Or a more verbose version of it if you prefer. If you use a shorter variable name for `difflist` (`dlist` maybe?) it may fit in single line statements.

File: libyang/data.py [hunk 6]
```diff
@@ -224,6 +281,31 @@ def validate(
...
+            yield DDiff(dtype, first, second)
+            i += 1
+
+        lib.lyd_free_diff(difflist)
```
Comment (rjarry) [inline libyang/data.py:307]: This will not be called if there is an error or if the generator is not consumed completely. You should wrap it with a `try`...`finally` block to make sure the allocated diff list is freed.

**Review by rjarry — CHANGES_REQUESTED** [review]
Hi @annikahannig,

I have made a few inline comments. Could you address them?

Also, could you squash all your commits into one with a nice commit log?

Thanks for your work!

File: libyang/data.py (hunk 5 above)
Comment (annikahannig) [reply libyang/data.py:303]: I was not sure if a <cdata 'void *' NULL> is falsy. It is.

Comment (annikahannig) [reply libyang/data.py]: yeah looked even better and more comapact after the change! However black made it worse. I hate this tool so much.

File: libyang/data.py [hunk 7]
```diff
@@ -224,6 +281,31 @@ def validate(
...
+
+        i = 0
+        try:
+            while dlist.type[i] != lib.LYD_DIFF_END:
+                first = (
+                    self.new(self.context, dlist.first[i]) if dlist.first[i] else None
+                )
+                second = (
+                    self.new(self.context, dlist.second[i]) if dlist.second[i] else None
+                )
```
Comment (rjarry) [inline libyang/data.py:303]: Indeed, it is not pretty...

Maybe use a verbose `if`...`else` instead?

Comment (annikahannig) [reply libyang/data.py:303]: Feel free to change.

Comment (annikahannig) [reply libyang/data.py]: I really hate this formatter.

Comment (annikahannig) [reply libyang/data.py]: I changed it again. looks better now.

**Comment by rjarry** [comment]
@annikahannig looks better indeed. For consistency with other commit titles, could you rephrase it like this?

```
data: add diff support

<short paragraph to explain what you have done>

<signed-off: yourself>
```

Thanks.

**Review by rjarry — APPROVED** [review]

**Comment by rjarry** [comment]
Thanks Annika :+1:

**Comment by rjarry** [comment]
I'll make a release with this (along with other fixes) today.

**Comment by annikahannig** [comment]
Cool! Thank you! :)
